src/platform/Linux/MdnsImpl.cpp

```diff
@@ -0,0 +1,665 @@
+    std::vector<TextEntry> textEntries;
+
+    switch (event)
+    {
+    case AVAHI_RESOLVER_FAILURE:
+        ChipLogError(DeviceLayer, "Avahi resolve failed");
+        context->mCallback(context->mContext, nullptr, CHIP_ERROR_INTERNAL);
+    case AVAHI_RESOLVER_FOUND:
+        MdnsResolveResult result;
+
+        ChipLogError(DeviceLayer, "Avahi resolve found");
+        strncpy(result.mService.mName, name, sizeof(result.mService.mName));
+        strncpy(result.mService.mType, type, sizeof(result.mService.mType));
+        result.mService.mName[kMdnsNameMaxSize] = 0;
+        result.mService.mType[kMdnsTypeMaxSize] = 0;
+        result.mService.mProtocol               = TruncateProtocolInType(result.mService.mType);
+        result.mService.mPort                   = port;
+
+        if (address)
+        {
+            switch (address->proto)
+            {
+            case AVAHI_PROTO_INET:
+                struct in_addr addr4;
+
+                memcpy(&addr4, &(address->data.ipv4), sizeof(addr4));
+                result.mAddress = chip::Inet::IPAddress::FromIPv4(addr4);
+            case AVAHI_PROTO_INET6:
+                struct in6_addr addr6;
+
+                memcpy(&addr6, &(address->data.ipv6), sizeof(addr6));
+                result.mAddress = chip::Inet::IPAddress::FromIPv6(addr6);
+            default:
+                break;
+            }
+        }
+
+        while (txt != nullptr)
+        {
+            TextEntry entry;
+
+            for (size_t i = 0; i < txt->size; i++)
+            {
+                if (txt->text[i] == '=')
+                {
+                    txt->text[i] = '\0';
+                    textEntries.push_back(TextEntry{ reinterpret_cast<char *>(txt->text), &txt->text[i + 1], txt->size - i - 1 });
+                    break;
+                }
+            }
+            txt = txt->next;
+        }
+
+        if (!textEntries.empty())
+        {
+            result.mService.mTextEntryies = textEntries.data();
+        }
+        result.mService.mTextEntrySize = textEntries.size();
+
+        context->mCallback(context->mContext, &result, CHIP_NO_ERROR);
+    }
+
+    avahi_service_resolver_free(resolver);
+}
+
+MdnsAvahi::~MdnsAvahi()
+{
+    if (mGroup)
+    {
+        avahi_entry_group_free(mGroup);
+    }
+    if (mClient)
+    {
+        avahi_client_free(mClient);
+    }
+}
+
+void UpdateMdnsDataset(fd_set & readFdSet, fd_set & writeFdSet, fd_set & errorFdSet, int & maxFd, timeval & timeout)
+{
+    MdnsAvahi::GetInstance().GetPoller().UpdateFdSet(readFdSet, writeFdSet, errorFdSet, maxFd, timeout);
+}
+
+void ProcessMdns(fd_set & readFdSet, fd_set & writeFdSet, fd_set & errorFdSet)
+{
+    MdnsAvahi::GetInstance().GetPoller().Process(readFdSet, writeFdSet, errorFdSet);
+}
+
+CHIP_ERROR ChipMdnsInit(MdnsAsnycReturnCallback initCallback, MdnsAsnycReturnCallback errorCallback, void * context)
+{
+    return MdnsAvahi::GetInstance().Init(initCallback, errorCallback, context);
```

Comment:
Can we use the context ptr to store the instance pointer?

Reply:
Done